sbi/diagnostics/sbc.py

```diff
@@ -0,0 +1,221 @@
+
+    Returns
+        ranks: ranks of true parameters vs. posterior samples under the specified RV,
+            for each posterior dimension.
+        log_prob_thos: log prob of true parameters under the approximate posterior.
+        dap_samples: samples from the data averaged posterior for the current batch,
+            i.e., a single sample from each approximate posterior.
+    """
+
+    log_prob_thos = torch.zeros(thos.shape[0])
+    dap_samples = torch.zeros_like(thos)
+    ranks = torch.zeros_like(thos)
+
+    for idx, (tho, xo) in enumerate(zip(thos, xos)):
+        # Log prob of true params under posterior.
+        log_prob_thos[idx] = posterior.log_prob(tho, x=xo)
+
+        # Rank true params vs posterior params under ranking_rv.
+        ths = posterior.sample((L,), x=xo, show_progress_bars=False)
+        dap_samples[idx] = ths[
+            0,
+        ]
+        # rank for each posterior dimension.
+        for dim in range(thos.shape[1]):
+            ranks[idx, dim] = (
+                (ranking_rv.log_prob(ths[:, dim]) < ranking_rv.log_prob(tho[dim]))
+                .sum()
+                .item()
+            )
+
+    return ranks, log_prob_thos, dap_samples
+
+
+def sbc_checks(
+    ranks: Tensor,
+    log_probs: Tensor,
+    prior_samples: Tensor,
+    dap_samples: Tensor,
+    num_ranks: int,
+):
+    """Return uniformity checks, data averaged posterior checks and NLTP for SBC."""
+    if ranks.shape[0] < 100:
+        warnings.warn(
+            """You are computing SBC checks with less than 100 samples. These checks
+            should be based on a large number of test samples theta_o, x_o. We
+            recommend using at least 100."""
+        )
+
+    ks_pvals, c2st_ranks = check_uniformity(ranks, num_ranks)
+    c2st_scrores_dap = check_prior_vs_dap(prior_samples, dap_samples)
+    nltp = torch.mean(-log_probs)
+
+    return dict(
+        ks_pvals=ks_pvals,
+        c2st_ranks=c2st_ranks,
+        c2st_dap=c2st_scrores_dap,
+        nltp=nltp,
+    )
+
+
+def check_prior_vs_dap(prior_samples: Tensor, dap_samples: Tensor):
+    """Return the c2st accuracy between prior and data avaraged posterior samples.
+
+    c2st is calculated for each dimension separately.
+    """
+
+    assert prior_samples.shape == dap_samples.shape
+
+    return torch.tensor(
+        [
+            c2st(s1.unsqueeze(1), s2.unsqueeze(1))
+            for s1, s2 in zip(prior_samples.T, dap_samples.T)
+        ]
+    )
+
+
+def check_uniformity(ranks, num_ranks, num_repetitions: int = 1):
+    """Return p-values and c2st scores for uniformity of the ranks.
+
+    Calculates Kolomogorov-Smirnov test using scipy.
+    """
+
+    from scipy.stats import kstest, uniform
+
+    kstest_pvals = torch.tensor(
+        [kstest(rks, uniform(loc=0, scale=num_ranks).cdf)[1] for rks in ranks.T],
+        dtype=torch.float32,
+    )
+
+    c2st_scores = torch.tensor(
+        [
+            [
+                c2st(
```

Comment:
Same as above, `c2st` uses a MLP to classify samples from both distributions, `rks` and `Uniform`. if ranks.shape[0] becomes large (in my experiments any value larger than `3` at say 10k samples), this function becomes really slooooow.

Reply:
thanks, that's a good point to address. let's make it an issue and link that to your timely PR #503